onnxruntime/core/framework/execution_context.h

```diff
@@ -0,0 +1,173 @@
+#pragma once
+#include "core/common/logging/logging.h"
+#include "core/framework/device_stream_collection.h"
+#include "core/framework/execution_frame.h"
+#include "core/framework/ort_value.h"
+#include "core/framework/iexecutor.h"
+#include "core/framework/stream_handles.h"
+#include "core/graph/basic_types.h"
+#include "core/common/inlined_containers.h"
+#include "core/framework/memory_info.h"
+#ifdef ENABLE_TRAINING
+#include "core/framework/partial_graph_execution_state.h"
+#endif
+
+namespace onnxruntime {
+class SessionState;
+
+class SessionScope;
+typedef InlinedHashMap<std::string, OrtValue> OrtValueCache;
+typedef std::shared_ptr<OrtValueCache> OrtValueCachePtr;
+
+// Execution context that support to execute a command on stream.
+// It is composed by following components:
+// 1. a execution frame
+// 2. a collection of device stream instances that kernels can launch to.
+// 3. a set of notification instances needed to perform synchronization in current execution plan.
+class ExecutionContext {
+ public:
+  /*
+   * LIMITATION:
+   * CountDownBarrier is only for scenario that the v is set
+   * to the # of consumers and each consumer calls Dec() exactly once.
+   */
+  class CountDownBarrier {
```

Comment:
1. Would waitable counter from nsync work instead of reinventing the wheel? https://github.com/google/nsync/blob/master/public/nsync_counter.h
2. If we've to write this, would be nice to keep it in a general utility folder/header file. core/util

Reply:
yes nsync counter will work, but nsync only available on non-Windows platforms, right? This is a relatively simple implementation, so i'd prefer just to implement it instead of use nsync on linux but another implementation on windows.
I am ok to keep it in separate file if we believe it is useful, once we confirm there is no other choice, i can work on separate it.

Reply:
let's continue the discussion in the new PR: #13495